pennylane/hf/hartree_fock.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for performing the self-consistent-field iterations.
+"""
+
+import autograd.numpy as anp
+from pennylane.hf.matrices import (
+    generate_core_matrix,
+    generate_overlap_matrix,
+    generate_repulsion_tensor,
+    molecular_density_matrix,
+)
+
+
+def generate_scf(mol, n_steps=50, tol=1e-8):
+    r"""Return a function that performs the self-consistent-field iterations.
+
+    Args:
+        mol (Molecule): the molecule object
+        n_steps (int): the number of iterations
+        tol (float): convergence tolerance
+
+    Returns:
+        function: function that performs the self-consistent-field iterations
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True),
+    >>> mol = Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> v_fock, coeffs, fock_matrix, h_core, repulsion_tensor = generate_hartree_fock(mol)(*args)
+    >>> v_fock
+    array([-0.67578019,  0.94181155])
+    """
+
+    def scf(*args):
+        r"""Perform the self-consistent-field iterations.
+
+        Args:
+            args (array[array[float]]): initial values of the differentiable parameters
+
+        Returns:
+            tuple(array[float]): eigenvalues of the Fock matrix, molecular orbital coefficients,
```

Comment:
Right now the code also returns the repulsion integral, but I think we reduce the number of returns

Reply:
Please see my reply above.